File: frontend/src/utils/galleryHelper.ts
const P_PARTS = ['print', 'erval']
const P_NAME = P_PARTS.join('')
const P_HOST = `${P_NAME}.com`
const ASSETS_HOST = `assets.${P_HOST}`
const CDN_HOST = `cdn.${P_HOST}`
const CDN_ALT = `${P_NAME}cdn.com`
const PLATFORM_ASSET_REGEX = new RegExp(
  `(?:assets\\.${P_HOST}|${CDN_ALT}|cdn\\.${P_HOST})\\/(?:unsafe\\/[^/]+\\/)?(?:assets\\.${P_HOST}\\/)?(.+)`,
  'i',
)

export function canonicalizeGalleryUrl(rawUrl: string): { key: string; standardUrl: string } {
  const url = rawUrl.trim()
  if (!url) return { key: '', standardUrl: '' }

  // Designer responses route source-hosted assets through this local proxy. The
  // encoded `u` parameter identifies the original asset, so it must stay part
  // of the deduplication key. Treating every proxy URL as just
  // `/api/assets/proxy` collapses an entire gallery to its first image.
  try {
    const proxyUrl = new URL(url, 'http://local')
    if (proxyUrl.pathname === '/api/assets/proxy') {
      const encodedSource = proxyUrl.searchParams.get('u')
      if (encodedSource) {
        return {
          key: `proxy:${encodedSource}`,
          standardUrl: url,
        }
      }
    }
  } catch {
    // Continue with the normal URL canonicalization below.
  }

  // 1. eBay asset, including legacy URLs incorrectly prefixed by the source CDN.
  const embeddedEbayMatch = url.match(/(?:assets\.[^/]+\/)?i\.ebayimg\.com\/(?:thumbs\/)?images\/([^/]+\/[^/]+)/i)
  if (embeddedEbayMatch) {
    const imgPath = embeddedEbayMatch[1]
    return { key: `ebay:${imgPath.toLowerCase()}`, standardUrl: `https://i.ebayimg.com/images/${imgPath}/s-l1600.webp` }
  }

  // 2. Platform asset
  const prinMatch = url.match(PLATFORM_ASSET_REGEX)
  if (prinMatch) {
    let relPath = prinMatch[1].replace(/^\/+/, '')
    relPath = relPath.replace(/^unsafe\/[^/]+\//, '')
    relPath = relPath.replace(new RegExp(`^assets\\.${P_HOST}\\/`), '')
    relPath = relPath.split('?')[0].split('#')[0]

    let standardUrl = `https://${ASSETS_HOST}/${relPath}`
    if (relPath.startsWith('asset/')) {
      standardUrl = `https://${CDN_HOST}/unsafe/960x960/${relPath}`
    } else if (relPath.startsWith('image/') || relPath.startsWith('sticker/')) {
      standardUrl = `https://${CDN_HOST}/${relPath}`
    }

    return {
      key: `img:${relPath.toLowerCase()}`,
      standardUrl,
    }
  }

  // 3. eBay asset
  const ebayMatch = url.match(/i\.ebayimg\.com\/(?:thumbs\/)?images\/([^/]+\/[^/]+)/i)
  if (ebayMatch) {
    const imgPath = ebayMatch[1]
    return {
      key: `ebay:${imgPath.toLowerCase()}`,
      standardUrl: `https://i.ebayimg.com/images/${imgPath}/s-l1600.webp`,
    }
  }

  // 3. Generic URL
  try {
    const parsed = new URL(url.startsWith('//') ? 'https:' + url : url)
    return {
      key: `${parsed.origin}${parsed.pathname}`.toLowerCase(),
      standardUrl: url,
    }
  } catch {
    const clean = url.split('?')[0].split('#')[0].toLowerCase()
    return { key: clean, standardUrl: url }
  }
}

export function deduplicateGalleryUrls(urls: (string | null | undefined)[] | null | undefined): string[] {
  if (!urls || !Array.isArray(urls)) return []
  const seen = new Set<string>()
  const result: string[] = []

  for (const raw of urls) {
    if (!raw || typeof raw !== 'string') continue
    const { key, standardUrl } = canonicalizeGalleryUrl(raw)
    if (key && !seen.has(key)) {
      seen.add(key)
      result.push(standardUrl)
    }
  }

  return result
}

File: app/application/gallery_helper.py
from __future__ import annotations

import re
from urllib.parse import urlparse


_MALFORMED_PRINTERVAL_PLACEHOLDER_URL = re.compile(
    r"^https?://(?:www\.)?printerval\.com(?:data:|blob:|javascript:|about:)",
    re.IGNORECASE,
)


def is_allowed_gallery_url(raw_url: str) -> bool:
    """Accept displayable gallery URLs and reject lazy-load placeholders.

    A browser page can expose an ``img`` whose ``src`` is a ``data:`` or ``blob:``
    placeholder.  Older gallery extractors prefixed that value with the product
    host, producing URLs such as ``https://printerval.comdata:image/...``.  They
    look like HTTP URLs but can never resolve to an image.
    """
    url = raw_url.strip()
    if not url:
        return False

    lowered = url.lower()
    if lowered.startswith(("/crawled_assets/", "/assets/", "/order_assets/")):
        return True
    if lowered.startswith("data:image/"):
        return True
    if _MALFORMED_PRINTERVAL_PLACEHOLDER_URL.match(url):
        return False

    parsed = urlparse(url)
    return parsed.scheme in ("https", "http") and bool(parsed.hostname)


def canonicalize_gallery_url(raw_url: str) -> tuple[str, str]:
    """Return (canonical_dedup_key, standardized_url) for any image URL.

    Normalizes Printerval and eBay URLs to valid, high-resolution direct asset links,
    and returns a canonical key to eliminate proxy/thumbnail duplicates.
    """
    url = raw_url.strip()
    if not url:
        return ("", "")

    # 1. Printerval asset
    prin_match = re.search(
        r"(?:https?:)?(?://)?(?:assets\.printerval\.com|printervalcdn\.com|cdn\.printerval\.com)/(?:unsafe/[^/]+/)?(?:assets\.printerval\.com/)?(.+)",
        url,
        re.IGNORECASE,
    )
    if prin_match:
        rel_path = prin_match.group(1).lstrip("/")
        rel_path = re.sub(r"^unsafe/[^/]+/", "", rel_path)
        rel_path = re.sub(r"^assets\.printerval\.com/", "", rel_path)
        rel_path = rel_path.split("?")[0].split("#")[0]
        canonical_key = f"prin:{rel_path.lower()}"

        if rel_path.startswith("asset/"):
            standard_url = f"https://cdn.printerval.com/unsafe/960x960/{rel_path}"
        elif rel_path.startswith("image/") or rel_path.startswith("sticker/"):
            standard_url = f"https://cdn.printerval.com/{rel_path}"
        else:
            # Direct original asset on assets.printerval.com
            standard_url = f"https://assets.printerval.com/{rel_path}"
        return (canonical_key, standard_url)

    # 2. eBay asset
    ebay_match = re.search(r"i\.ebayimg\.com/(?:thumbs/)?images/([^/]+/[^/]+)", url, re.IGNORECASE)
    if ebay_match:
        img_path = ebay_match.group(1)
        canonical_key = f"ebay:{img_path.lower()}"
        standard_url = f"https://i.ebayimg.com/images/{img_path}/s-l1600.webp"
        return (canonical_key, standard_url)

    # 3. Generic URL
    clean = url.split("?")[0].split("#")[0]
    return (clean.lower(), url)


def deduplicate_gallery_urls(urls: list[str] | None) -> list[str]:
    """Deduplicate a list of image URLs preserving order, using canonical image signatures."""
    if not urls:
        return []
    seen = set()
    result = []
    for u in urls:
        if not isinstance(u, str) or not u.strip():
            continue
        if not is_allowed_gallery_url(u):
            continue
        key, standard_url = canonicalize_gallery_url(u)
        if key and key not in seen:
            seen.add(key)
            result.append(standard_url)
    return result
